### PR TITLE
[Console] add setInputs to ApplicationTester and share some code

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes
| Fixed tickets | #24784
| License       | MIT
| Doc PR        | todo

I didn't implemented the tests because I don't know how to write them on ApplicationTester.

### CODE REVIEW

File: src/Symfony/Component/Console/Tester/TesterTrait.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * Sets the user inputs.
+     *
+     * @param $inputs array an array of strings representing each input
+     *              passed to the command input stream
+     *
+     * @return self
+     */
+    public function setInputs(array $inputs)
```
Comment (chalasr) [inline src/Symfony/Component/Console/Tester/TesterTrait.php]: For this to work in ApplicationTester, the tester needs to set the input stream from this value.
So `CommandTester::createInputStream()` should be shared between both testers (that's my motivation for a trait) and [this line](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/Tester/CommandTester.php#L65..L67) should be added to `ApplicationTester::run()`.

File: src/Symfony/Component/Console/Tester/TesterTrait.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        $this->inputs = $inputs;
```
Comment (chalasr) [inline src/Symfony/Component/Console/Tester/TesterTrait.php]: the property does not exist in ApplicationTester, it should be created

**Review by chalasr — CHANGES_REQUESTED** [review]
here is a test case
```patch
diff --git a/src/Symfony/Component/Console/Tests/Tester/ApplicationTesterTest.php b/src/Symfony/Component/Console/Tests/Tester/ApplicationTesterTest.php
index 062d414..0cc002d 100644
--- a/src/Symfony/Component/Console/Tests/Tester/ApplicationTesterTest.php
+++ b/src/Symfony/Component/Console/Tests/Tester/ApplicationTesterTest.php
@@ -13,7 +13,9 @@ namespace Symfony\Component\Console\Tests\Tester;
 
 use PHPUnit\Framework\TestCase;
 use Symfony\Component\Console\Application;
+use Symfony\Component\Console\Helper\QuestionHelper;
 use Symfony\Component\Console\Output\Output;
+use Symfony\Component\Console\Question\Question;
 use Symfony\Component\Console\Tester\ApplicationTester;
 
 class ApplicationTesterTest extends TestCase
@@ -65,6 +67,22 @@ class ApplicationTesterTest extends TestCase
         $this->assertEquals('foo'.PHP_EOL, $this->tester->getDisplay(), '->getDisplay() returns the display of the last execution');
     }
 
+    public function testSetInputs()
+    {
+        $this->application->register('foo')->setCode(function ($input, $output) {
+            $helper = new QuestionHelper();
+            $helper->ask($input, $output, new Question('Q1'));
+            $helper->ask($input, $output, new Question('Q2'));
+            $helper->ask($input, $output, new Question('Q3'));
+        });
+
+        $this->tester->setInputs(array('I1', 'I2', 'I3'));
+        $this->tester->run(array('command' => 'foo'));
+
+        $this->assertSame(0, $this->tester->getStatusCode());
+        $this->assertEquals('Q1Q2Q3', $this->tester->getDisplay(true));
+    }
+
     public function testGetStatusCode()
     {
         $this->assertSame(0, $this->tester->getStatusCode(), '->getStatusCode() returns the status code');

```

File: src/Symfony/Component/Console/Tester/TesterTrait.php (hunk 2 above)
Comment (ogizanagi) [reply src/Symfony/Component/Console/Tester/TesterTrait.php]: or moved to the trait along with:

```php
    /**
     * @var StreamOutput
     */
    private $output;
```

**Comment by Simperfit** [comment]
@chalasr So it seems that the test is failling. Is here something the modify directly in the Application to make this work properly ?

File: src/Symfony/Component/Console/Tester/TesterTrait.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+use Symfony\Component\Console\Output\StreamOutput;
+
+/**
+ * @author Amrouche Hamza <hamza.simperfit@gmail.com>
+ *
+ * @internal
+ */
+trait TesterTrait
+{
+    /** @var StreamOutput */
```
Comment (chalasr) [inline src/Symfony/Component/Console/Tester/TesterTrait.php]: not useful and differs with `getOutput()` return type, I would drop it

File: src/Symfony/Component/Console/Tester/ApplicationTester.php [hunk 4]
```diff
@@ -62,10 +58,15 @@ public function __construct(Application $application)
     public function run(array $input, $options = array())
     {
         $this->input = new ArrayInput($input);
+
         if (isset($options['interactive'])) {
             $this->input->setInteractive($options['interactive']);
         }
 
+        if ($this->inputs) {
+            $this->input->setStream(self::createStream($this->inputs));
```
Comment (chalasr) [inline src/Symfony/Component/Console/Tester/ApplicationTester.php]: [Here is the failure culprit](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/Application.php#L787), I suggest to `putenv('SHELL_INTERACTIVE=1')` here and, after `run()`, either destroy it or restore its previous value if any

File: src/Symfony/Component/Console/Tester/TesterTrait.php (hunk 3 above)
Comment (ogizanagi) [reply src/Symfony/Component/Console/Tester/TesterTrait.php]: To me it is useful actually (for autocompletion and STA). This trait needs a `StreamOutput` instance here, not any `OutputInterface`. That's why I suggested this typehint in https://github.com/symfony/symfony/pull/24819#discussion_r148958291 :)

**Comment by Simperfit** [comment]
Thanks for the review guys, tests are now green ;).

File: src/Symfony/Component/Console/Tester/TesterTrait.php [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    private $output;
+    private $inputs = array();
+
+    /**
+     * Gets the display returned by the last execution of the application.
```
Comment (chalasr) [inline src/Symfony/Component/Console/Tester/TesterTrait.php]: `command or application`, same below

File: src/Symfony/Component/Console/Tester/TesterTrait.php [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+     * @return self
+     */
+    public function setInputs(array $inputs)
+    {
+        $this->inputs = $inputs;
+
+        return $this;
+    }
+
+    public static function createStream(array $inputs)
```
Comment (chalasr) [inline src/Symfony/Component/Console/Tester/TesterTrait.php]: this one must stay private

File: src/Symfony/Component/Console/Tester/ApplicationTester.php [hunk 7]
```diff
@@ -97,27 +101,15 @@ public function run(array $input, $options = array())
...
-     * @param bool $normalize Whether to normalize end of lines to \n or not
-     *
-     * @return string The display
-     */
-    public function getDisplay($normalize = false)
-    {
-        rewind($this->output->getStream());
-
-        $display = stream_get_contents($this->output->getStream());
+        putenv('SHELL_INTERACTIVE');
```
Comment (chalasr) [inline src/Symfony/Component/Console/Tester/ApplicationTester.php]: `putenv($shellInteractive ? "SHELL_INTERACTIVE=$shellInteractive" : 'SHELL_INTERACTIVE');` instead, removing the condition below

**Review by chalasr — APPROVED** [review]
👍

File: src/Symfony/Component/Console/Tester/ApplicationTester.php [hunk 8]
```diff
@@ -62,10 +58,18 @@ public function __construct(Application $application)
     public function run(array $input, $options = array())
     {
         $this->input = new ArrayInput($input);
+
```
Comment (ogizanagi) [inline src/Symfony/Component/Console/Tester/ApplicationTester.php]: nitpicking but should be reverted ^^

**Review by ogizanagi — APPROVED** [review]

**Review by 20uf — APPROVED** [review]

**Comment by chalasr** [comment]
Thank you @Simperfit.

**Pull request merged**